sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryHelper.scala

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.jni.{RetryOOM, RmmSpark, SplitAndRetryOOM}
+
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+/**
+ * RmmRapidsRetryHelper is used to attempt work on the GPU with the ability to retry it,
+ * abstracting the retry logic from the calling code.
+ * @param input a SpillableColumnarBatch that we would like to attempt work with
+ * @param splitPolicy a function that can split a SpillableColumnarBatch into multiple
+ *                    spillable batches, can be set to null in case splits are not handled.
+ * @param mergePolicy a function that can be used to merge SpillableColumnarBatches that had
+ *                    been previously split, can be set to null in case merging is not handled
+ *                    or not required.
+ * @param maxRetries the maximum amount of times the retry loop will re-attempt on RetryOOM or
+ *                   SplitAndRetryOOM (3 retries by default)
+ */
```

Comment:
That would be great. It seems that it gives the retry state machine a lot more flexiblity too. It may decide not to retry anymore for example, so a static number of retries may not be a good thing. @revans2 what do you think?

Reply:
The problem with putting it into the centralized logic is there is not feedback that this task made progress.  We have talked about adding in an API that is called when this section is done so it knows to reset the counter. That would totally be doable. If that is what we want just let me know and I'll add it.

But either way I think this needs to come from a config. I don't see why we would want different operations to have a different number of retries.

Reply:
I was under the impression the state machine was going to handle retry vs. OOM and there wasn't any concept of a retry limit.  There's retrying because we need this thread to clear memory so other threads can make progress, but if we get a low-level alloc failure and this is the only thread, there's going to be a split-retry exception thrown.  If the app code can no longer split (input is too small or whatever) then the app will OOM at that point.  I don't see a need for some arbitrary retry count, even at a low level.  We keep splitting and retrying until we can't split further.

Reply:
For now I am removing numRetries/maxRetries from here assuming that there will be a different exception thrown other then the retry exceptions. It's a minor change to add it back later if we decide we want some sort of watchdog at this level.